### DISCUSSION TITLE
Making curl impersonate Firefox

### DISCUSSION THREAD

**Discussion by emanruse** [curl/curl#15216] [Ideas]
Hi,

I have been trying to download RSS feeds of some websites using curl behind a Tor proxy. Unfortunately, even if I use the identical headers which I see in Tor Browser's console, the remote host blocks me or represents a JS challenge, thus making it impossible to access the content. The result is - the same address **can** be accessed in Tor Browser but **not** through curl behind Tor. Example:

```
curl \
	--http2 \
	--proxy socks5h://${RANDOM}:${RANDOM}@127.0.0.1:9050 \
	-A 'Mozilla/5.0 (Windows NT 10.0; rv:109.0) Gecko/20100101 Firefox/115.0' \
	-H 'Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8' \
	-H 'Accept-Encoding: gzip, deflate, br' \
	-H 'Accept-Language: en-US,en;q=0.5' \
	-H 'Upgrade-Insecure-Requests: 1' \
	-H 'Cache-Control: no-cache' \
	-H 'Pragma: no-cache' \
	-H 'Sec-Fetch-Dest: document' \
	-H 'Sec-Fetch-Mode: navigate' \
	-H 'Sec-Fetch-Site: none' \
	-H 'Sec-Fetch-User: ?1' \
	-H 'TE: trailers' \
	--location \
	--max-redirs 5 \
	--junk-session-cookies \
	--tlsv1.2 \
	--ciphers NULL-SHA256 \
	--tls13-ciphers TLS_AES_128_GCM_SHA256:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_256_GCM_SHA384:TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256:TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256:TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256:TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256:TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384:TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384:TLS_RSA_WITH_AES_128_GCM_SHA256:TLS_RSA_WITH_AES_256_GCM_SHA384 \
	--compressed \
	https://www.canonrumors.com/feed/
```

This is obviously quite unreasonable, because it should be perfectly fine to download such text content while staying anonymous - there is zero harm to the website. However, popular protection proxies like Cloudflare stimulate their users to be overprotective, thus blocking legit requests.

While looking for a solution, I found [this article](https://lwthiker.com/reversing/2022/02/17/curl-impersonate-firefox.html) and the [curl-impersonate](https://github.com/lwthiker/curl-impersonate) project. I built it from source and then I could `curl_ff117 https://www.canonrumors.com/feed/` from a torified VM which worked fine. The problem is that:

- curl-impersonate does not provide e.g. Tor Browser-specific config, which is my particular goal
- it is inconvenient to rely on yet another non-distro packaged tool and build it from source

So, my question is:

Is it possible for curl devs to look into this and possibly provide additional CLI options, so that one can impersonate any browser, similar to curl-impersonate? If yes, how should that be requested?

**Comment by dfandrich** [comment]
I don't think you'll find much enthusiasm here for starting an arms race with the CDNs and bit web properties. Adding features to the standard curl to do this just means they'll be working harder to work against those features and meanwhile, the curl code become harder to work with. curl intends to follow the RFCs to perform transfers, and it does that very well. The kinds of things that would need to be added to curl to impersonate browsers are unnecessary for those aims.

If people running servers want to explicitly prevent their users from using curl to access them, they are within their rights to do that (generally speaking) and it's not curl's job to try to work around the measures they have put into place to legitimately control access to their sites. You'll also probably find that many of those same sites have provisions in their terms of service that prevent access using the kinds of programs that often use curl, so trying to overcome those restrictions can in some jurisdictions, actually open you up to legal criminations.

**Comment by emanruse** [comment]
Thanks for the feedback.

The first paragraph of your reply makes sense but browsers also follow RFCs, so impersonation won't go against that.

> If people running servers want to explicitly prevent their users from using curl to access them, they are within their rights to do that (generally speaking) and it's not curl's job to try to work around the measures they have put into place to legitimately control access to their sites.

Well, people do not generally want to prevent users from using curl. The actual goal they have is to prevent mass scraping and that is OK. However, to the best of my knowledge, most (if not all) RSS readers use exactly (lib)curl to fetch content. In that sense, there is nothing illegal in that. The problem is that a great deal of people who own websites simply don't understand that and just "play it safe" by blocking/challenging all "non-browser" HTTP requests for all URLs. The result: RSS simply becomes dysfunctional as no RSS reader will solve JS challenges.

My point is: The fact that most website owners (e.g. WordPress publishers) don't even know what curl is and what I explained above is to their detriment too. In that sense, having the extra flexibility of an impersonation functionality would benefit both sides.

Yes, it can be misused but if one has that goal, he will find ways. That's not what I am suggesting here.

I hope that clarifies.

**Comment by bagder** [comment]
Lots of places these days do *TLS fingerprinting* as a way to try to make sure that the client is a known and "valid" browser. I blogged about it two years ago: https://daniel.haxx.se/blog/2022/09/02/curls-tls-fingerprint/

That's also the area for which curl-impersonate has the most work and adaptions (at least that was the case when I checked it out, although admittedly that was a while ago): it needs to do TLS **exactly** as the intended browser so that it appears as one of those in the TLS handshake.

**Reply by perkfly** [reply]
One of the missing pieces is Pre-Shared Keys. In my observation, browsers always use the PSK after the handshake, while curl does not support PSK at the moment.

**Reply by MasterInQuestion** [reply]
Session resumption with PSK maybe deliberately restricted for privacy considerations.
    So not a reliable factor as fingerprinting.
    [ Also session resumption is inapplicable for 1st access. ]

    And actually there are not many factors that may be relatively reliably used to discern the traffic originator.
    See also: [ https://github.com/MasterInQuestion/talk/discussions/20](https://github.com/MasterInQuestion/talk/discussions/20)

**Comment by emanruse** [comment]
> That's also the area for which curl-impersonate has the most work and adaptions (at least that was the case when I checked it out, although admittedly that was a while ago): it needs to do TLS **exactly** as the intended browser so that it appears as one of those in the TLS handshake.

Yes.

**Comment by MasterInQuestion** [comment]
"TLS_EMPTY_RENEGOTIATION_INFO_SCSV":
    [ https://github.com/curl/curl/issues/13873#issuecomment-2183960844](https://github.com/curl/curl/issues/13873#issuecomment-2183960844)

https://archive.org/details/4431785826199-Fix-Discord-Zendesk-Sign-In
    “The presence of "TLS_EMPTY_RENEGOTIATION_INFO_SCSV" (0xFF) in the supported TLS Cipher Suites appears to be the cause:
    Which selected by Cloudflare as the typically access pattern of non-browsers.”
